src/runtime/refiner.ts

```diff
@@ -8,142 +8,699 @@
+        }
+        if (a.to.val === b.to.val) {
+            right = {...a.to};
+            right.kind = a.to.kind === b.to.kind ? a.to.kind : 'closed';
+        } else {
+            right = a.to.val > b.to.val ? {...a.to} : {...b.to};
+        }
+        return new Segment(left, right);
+    }
+
+    static overlap(a: Segment, b: Segment): Segment {
+        if (!a.overlapsWith(b)) {
+            throw new Error('Cannot find intersection of non-overlapping segments');
+        }
+        let left: Boundary; let right: Boundary;
+        if (a.from.val === b.from.val) {
+            left = {...a.from};
+            left.kind = a.from.kind === b.from.kind ? a.from.kind : 'open';
+        } else {
+            left = a.from.val > b.from.val ? {...a.from} : {...b.from};
+        }
+        if (a.to.val === b.to.val) {
+            right = {...a.to};
+            right.kind = a.to.kind === b.to.kind ? a.to.kind : 'open';
+        } else {
+            right = a.to.val < b.to.val ? {...a.to} : {...b.to};
         }
+        return new Segment(left, right);
+    }
+}
+
+interface Boundary {
+    val: number;
+    kind: 'open' | 'closed';
+}
+
+// Does not contain == and !=
+const operandCompatability = {
+    'and': ['Boolean', 'Boolean'],
+    'or': ['Boolean', 'Boolean'],
+    '<': ['Number', 'Number'],
+    '>': ['Number', 'Number'],
+    '<=': ['Number', 'Number'],
+    '>=': ['Number', 'Number'],
+    '+': ['Number', 'Number'],
+    '-': ['Number', 'Number'],
+    '*': ['Number', 'Number'],
+    '/': ['Number', 'Number'],
+    'not': ['Boolean'],
+    'neg': ['Number'],
+};
+
+class RefinementOperator {
+    op: string;
+    fn: Evaluator;
+
+    constructor(operator: string) {
+        this.op = operator;
+        this.updateFn();
+    }
+
+    updateOp(op: string) {
+        this.op = op;
+        this.updateFn();
+    }
+
+    private updateFn() {
+        switch (this.op) {
+            case 'and': this.fn = (expr) => expr[0] && expr[1]; break;
+            case 'or': this.fn = (expr) => expr[0] || expr[1]; break;
+            case '<': this.fn = (expr) => expr[0] < expr[1]; break;
+            case '>': this.fn = (expr) => expr[0] > expr[1]; break;
+            case '<=': this.fn = (expr) => expr[0] <= expr[1]; break;
+            case '>=': this.fn = (expr) => expr[0] >= expr[1]; break;
+            case '+': this.fn = (expr) => expr[0] + expr[1]; break;
+            case '-': this.fn = (expr) => expr[0] - expr[1]; break;
+            case '*': this.fn = (expr) => expr[0] * expr[1]; break;
+            case '/': this.fn = (expr) => expr[0] / expr[1]; break;
+            case 'not': this.fn = (expr) => !expr[0]; break;
+            case 'neg': this.fn = (expr) => -expr[0]; break;
+            case '==': this.fn = (expr) => expr[0] === expr[1]; break;
+            case '!=': this.fn = (expr) => expr[0] !== expr[1]; break;
+            default: throw new Error(`Invalid refinement operator ${this.op}`);
```

Comment:
I would be inclined to fold this and the `evalType` switch below into the `operandCompatability` table (which would now hold all operators, including `==` and `!=`). That will simplify the class quite a lot:
```
interface OperatorInfo {
  nArgs: number;
  argType: string;
  evalType: string;
  evalFn: (exprs: ExpressionPrimitives[]) => ExpressionPrimitives;
}

const operandTable: Dictionary<OperatorInfo> = {
  'and': {nArgs: 2, argType: 'Boolean', evalType: 'Boolean', evalFn: e => e[0] && e[1]},
  'or':  {nArgs: 2, argType: 'Boolean', evalType: 'Boolean', evalFn: e => e[0] || e[1]},
  '<':   {nArgs: 2, argType: 'Number',  evalType: 'Boolean', evalFn: e => e[0] < e[1]},
  ...
  '==':  {nArgs: 2, argType: 'same',    evalType: 'Boolean', evalFn: e => e[0] === e[1],
  '!=':  {nArgs: 2, argType: 'same',    evalType: 'Boolean', evalFn: e => e[0] !== e[1],
}

class RefinementOperator {
  opInfo: OperatorInfo;

  constructor(operator: string) {
    this.updateOp(operator);
  }

  updateOp(operator: string) {
    this.opInfo = operandTable[operator];
    if (!this.opInfo) {
      throw new Error(`Invalid refinement operator ${operator}`);
    }
  }

  eval(exprs: ExpressionPrimitives[]): ExpressionPrimitives {
    return this.opInfo.evalFn(exprs);
  }

  evalType(): 'Number' | 'Boolean' {
    return this.opInfo.evalType;
  }

  assertOperandCompatibility(operandTypes: string[]): void {
    if (operandTypes.length !== this.opInfo.nArgs) {
      throw new Error(...);
    }
    if (this.opInfo.argType === 'same') {
      if (operandTypes[0] !== operandTypes[1]) {
        throw new Error(`Expected ${operandTypes[0]} and ${operandTypes[1]} to be the same.`);
      }
    } else {
      for (const type of operandTypes) {
        if (type !== this.opInfo.argType) {
          throw new Error(`Got type ${type}. Expected ${this.opInfo.argType}.`);
        }
      }
    }
  }
}
```